### PR TITLE
Fix forward declaration of basic_string and list, for dual abi libstdc++

### PR DESCRIPTION
**Pull request by xnox**
Fix forward declaration of basic_string and list, for dual abi libstdc++ with inline std::__cxx11 namespace.

Closes facebook/folly#213
Closes facebook/folly#208

### CODE REVIEW

**Comment by xnox** [comment]
This is an example on how to fix this problem and is trivial. There are other ways, e.g. detect that dual abi is available by querying  _GLIBCXX_USE_CXX11_ABI and if set to 1 properly forward declare inlined namespace std::* or skip forward declarations.

Can maintainers comment on this merge proposal? I don't see point to go through employeer cla, if this change is trivial / non-copyrightable / will not be merged / etc.

**Comment by sgolemon** [comment]
So this is trivial enough that yeah, we could probably forego the CLA, *but* it's related to another PR #81 and there's some concern from the main folly team about including the official headers where a handful of forward declarations would do...

I'm not convinced of their argument, but we're gonna fix the underlying problem regardless.

For now, I'm going to close this as a duplicate of #81 and what we land will probably be something different from either.

TL;DR - I wouldn't worry about the CLA :)

**Comment by xnox** [comment]
@sgolemon Thanks a lot. Forward declaration is a compile time optimisation, and indeed they are correct. However, libstdc++ 5.0 does something rather crazy, and the "C++11" abi declarations are wrapped in an inline namespace. Thus "std::string" is actually defined as "std::__cxx11::string" (with the __cxx11 inline namespace). I think the compiler is wrong here, and that a forward declaration should resolve to the inline namespace implementation, instead of resolving as a name conflict. Gcc people cite the holy standard, and pretend it is wrong to forward declare "std::*". If the forward declaration can be "inlined" to match __cxx11, that would work... but i have no idea how to do that, hence deployed the hammer solution of a full include. As I really don't want to get involved into the code-style battle of forward declares (i.e. if one should use them, if they should work with dual-abi libstdc++, and whose behaviour/expectations are correct).
